nft/src/main.sw

```diff
@@ -0,0 +1,488 @@
+contract;
+
+dep abi;
+dep errors;
+dep events;
+
+use abi::NFT;
+use errors::{AccessError, ApprovalError, InitError, InputError};
+use events::{ApprovalEvent, BurnEvent, MintEvent, OperatorEvent, TransferEvent};
+
+use std::{
+    address::Address,
+    assert::require,
+    chain::auth::{AuthError, msg_sender},
+    constants::NATIVE_ASSET_ID,
+    context::{call_frames::{contract_id, msg_asset_id}, msg_amount, this_balance},
+    contract_id::ContractId,
+    hash::sha256,
+    identity::Identity,
+    logging::log,
+    option::Option,
+    result::*,
+    revert::revert,
+    storage::StorageMap,
+};
+
+struct MetaData {
+    // NFT Metadata
+    owner: Identity,
+    approved: Identity,
+}
+
+storage {
+    /// Determines if there is a whitelist to mint. Only set on the
+    /// initalization of the contract
+    access_control: bool,
+
+    /// The identity that has permission to add identities to the whitelist
+    access_control_address: Identity,
+
+    /// Stores the identities that are on the whitelist
+    allowed_minters: StorageMap<Identity, bool>,
+
+    /// Maintains the number of tokens owned by each of the identities
+    balances: StorageMap<Identity, u64>,
+
+    /// The asset accepted as payment for minting a token
+    asset: ContractId,
+
+    /// The metadata for each token based on token id
+    meta_data: StorageMap<u64, MetaData>,
+
+    /// Stores a b256 hash of the (owner, operator) and stores whether the
+    /// operator is allowed to transfer ALL tokens on the owner's behalf 
+    operator_approval: StorageMap<b256, bool>,
+
+    /// TODO: Use a Vec here to support multiple ownerships
+    /// Maintains the token ids owned by the specified identity
+    owners: StorageMap<Identity, u64>,
+
+    /// The state of the contract. Can either be initalized or not
+    state: u64,
+
+    /// The total number of tokens that have been minted
+    token_count: u64,
+
+    /// The price to purchase a single token. Can only be set on
+    /// the initalization of the contract
+    token_price: u64,
```

Comment:
I don't think the general-purpose base contract needs to have a notion of token price.
Not all NFTs needs be for sale, and the pricing logic is easily added in custom implementation s that require it.

Reply:
To further that thought, even when the token has a price, the price may be a hardcoded constant, or may be provided by an oracle, or may be calculated using some flavour of bonding-curve, etc...